### PR TITLE
Support horovod

### PR DESCRIPTION
**Pull request by zuston**

### CODE REVIEW

**Comment by zuston** [comment]
This PR is just to support static horovod by using gloo controller. Elastic horovod will support in later PR.

**Comment by oliverhu** [comment]
Thanks for the PR!! Looks good at a glance, not very sure how Horovod works elastically, looking forward to the design doc! Please also add some integration tests, and examples.

**Comment by zuston** [comment]
Attached local test case.
https://github.com/zuston/horovod-yarn

**Comment by zuston** [comment]
> Thanks for the PR!! Looks good at a glance, not very sure how Horovod works elastically, looking forward to the design doc! Please also add some integration tests, and examples.

Example and proposal are all provided. Please check it.
More tests will be added in the few days. @oliverhu

**Comment by oliverhu** [comment]
@erwa  @goyalankit  could you also review this? We tried to resolve this before but didn't get a chance to finish.

File: docs/proposals/horovod-on-tony.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+1. Support Horovod on TonY (limited static topology support)
+2. Elastic Horovod will be supported later.
+
+## API
+There are no API changes to other machine learning frameworks.
+
+## Design
+Attention: It's gloo controller that makes TonY support horovod.
+
+From the perspective of compatibility and maintainability, it's better to directly use the [gloo_runner](https://github.com/horovod/horovod/blob/master/horovod/runner/gloo_run.py) on TonY. But after reading through Horovod's code, I find it difficult to reuse gloo_runner's code on TonY, because the existence of some unrelated codes will lead driver to start the worker through the SSH command, which is hard to be supportted on Yarn.
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:24]: Will there be any compatibility issue later if Horovod makes changes?

File: docs/proposals/horovod-on-tony.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+__Where to start driver__  
+Two options  
+1. On TonY application master.  
+This will save resources(no extra resources to start driver), and the amount of code changes will be small. But by injecting relevant Horovod's driver code into AM, it is not elegant.
+2. On TonY task executor.  
+Additional customization of the driver configuration is required and allow different task executors to specify different resources(because driver needn't gpu). And it is necessary to coordinate the startup sequence between the driver and other workers, because driver should start before worker. But 
+we should make some changes to prevent starting in each task executor at the same time.  
+
+First option will be adopted. The second one remains to be supported in the future.
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:59]: I prefer using #2 . 
- AM should only be used for container orchestration. We don't want any framework specific code to run there.
- TonY supports a mini DAG engine: https://github.com/linkedin/TonY/blob/master/tony-core/src/main/java/com/linkedin/tony/TonyConfigurationKeys.java#L131 , tasks in prepare-stage will finish first before training-stage starts.

File: docs/proposals/horovod-on-tony.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+HOROVOD_GLOO_RENDEZVOUS_PORT=9999
+HOROVOD_GLOO_RENDEZVOUS_ADDR=localhost
+
+HOROVOD_CROSS_RANK=0
+HOROVOD_CROSS_SIZE=1
+HOROVOD_LOCAL_RANK=0
+HOROVOD_LOCAL_SIZE=1
+HOROVOD_SIZE=1
+HOROVOD_RANK=0
+HOROVOD_HOSTNAME=192.168.0.1
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:82]: 0.0.0.0?

File: tony-examples/horovod-on-tony/tensorflow2_mnist.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+for batch, (images, labels) in enumerate(dataset.take(10000 // hvd.size())):
+    loss_value = training_step(images, labels, batch == 0)
+
+    if batch % 10 == 0 and hvd.local_rank() == 0:
+        print('Step #%d\tLoss: %.6f' % (batch, loss_value))
+
+# Horovod: save checkpoints only on worker 0 to prevent other workers from
+# corrupting it.
+if hvd.rank() == 0:
+    checkpoint.save(checkpoint_dir)
```
Comment (oliverhu) [inline tony-examples/horovod-on-tony/tensorflow2_mnist.py:101]: add a new line below

File: tony-core/src/test/resources/horovod_driver.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
+#
```
Comment (oliverhu) [inline tony-core/src/test/resources/horovod_driver.py:1]: why do you need to duplicate this file?

File: docs/proposals/horovod-on-tony.md (hunk 1 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:24]: Yes. 
If horovod changes the core api about rendezvous server, TonY needs to be compatible  with it.

File: docs/proposals/horovod-on-tony.md (hunk 2 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:59]: Good idea. It will work by using **prepare-stage**.

File: docs/proposals/horovod-on-tony.md (hunk 3 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:82]: OK

File: tony-core/src/test/resources/horovod_driver.py (hunk 5 above)
Comment (zuston) [reply tony-core/src/test/resources/horovod_driver.py:1]: Added it by mistake. I will remove it.

Comment (zuston) [reply tony-core/src/test/resources/horovod_driver.py]: Done

File: tony-examples/horovod-on-tony/tensorflow2_mnist.py (hunk 4 above)
Comment (zuston) [reply tony-examples/horovod-on-tony/tensorflow2_mnist.py]: Done

File: docs/proposals/horovod-on-tony.md (hunk 2 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:59]: It's hard to reuse prepare-stage conf, TonY will track and wait for to finish before starting tasks in the training stage.
However, we just need to keep startup sequence and ensure that the **driver** and **worker** life cycles are the same.

Comment (oliverhu) [reply docs/proposals/horovod-on-tony.md:59]: @UWFrankGu  how did we use `pre-stage` to load data into redis before starting the training job?

Comment (UWFrankGu) [reply docs/proposals/horovod-on-tony.md:59]: We added `redis` servers in `tony.application.untracked.jobtypes` to ensure that the `redis` process runs throughout `prepare-stage` and `training-stage`. More information in the TonY Configurations doc: https://github.com/linkedin/TonY/wiki/TonY-Configurations

Comment (oliverhu) [reply docs/proposals/horovod-on-tony.md:59]: ah right.. you also need `tony.application.untracked.jobtypes` 😂

**Comment by erwa** [comment]
> @erwa @goyalankit could you also review this? We tried to resolve this before but didn't get a chance to finish.

Saw you mentioned me. I don't think I'm the best to review given I haven't touched TonY code in almost two years lol. I'm not familiar with the new stages in TonY or the new stuff in Horovod (though it's cool to see we introduced new stages to support the Redis use case, which I remember well :).

One general comment, though, is that I see there is some refactoring work in this PR, too, such as the MLFrameworkRuntime refactoring and adding a lot of getters to TaskExecutor.java. I think the refactoring is nice (the MLFrameworkRuntime seems like a clean interface) but I think it should be done in a separate PR first to keep the PRs smaller and more discrete. Just a suggestion -- I'll let you guys who are actively working on/maintaining TonY make the call.

(P.S.: One small nit in the MLFrameworkRuntime interface: there's a typo: destory -> destroy)

File: docs/proposals/horovod-on-tony.md (hunk 2 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:59]: I think i will not use `stage` to coordinate driver and worker startup sequence.
Through extending the `canStart()` method in `MLFrameworkRuntime` interface, it's more elegant to coordinate each task executor's python process startup sequence.

**Comment by zuston** [comment]
> One general comment, though, is that I see there is some refactoring work in this PR, too, such as the MLFrameworkRuntime refactoring and adding a lot of getters to TaskExecutor.java. I think the refactoring is nice (the MLFrameworkRuntime seems like a clean interface) but I think it should be done in a separate PR first to keep the PRs smaller and more discrete.

A good suggestion! I will refactor it in another separate PR.

File: docs/proposals/horovod-on-tony.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+__Where to start driver__  
+Two options  
+1. On TonY application master.  
+This will save resources(no extra resources to start driver), and the amount of code changes will be small. But by injecting relevant Horovod's driver code into AM, it is not elegant.
+2. On TonY task executor.  
+Additional customization of the driver configuration is required and the startup of driver will be covered on TonY automatically. And it is necessary to coordinate the startup sequence between the driver and other workers, because driver should start before worker. 
+
+Second option will be adopted in this PR.  
+In order to unify different machine framework startup, we supposed to create `MLFrameworkRuntime` interface, it will expose methods as follows
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:59]: Just call it `FrameworkRuntime`, not limited to MLFramework?

File: docs/proposals/horovod-on-tony.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    /** For AM, getting cluster spec and return to task exectuor **/
+    String constructClusterSpec(String taskId) throws IOException;
+
+    /** For AM, when app finished, it need to call it to release resource **/
+    void destroy();
+
+    /** For AM, init the tony session **/
+    void setTonySession(final TonySession session);
+
+    /** For AM, it ensures that each task executor start sequence. like Horovod driver should start before workers **/
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:70]: add a bit more documentation.. it is not clear what can start. Maybe a differnet name for the function as well

File: docs/proposals/horovod-on-tony.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    boolean canStart(TonyConfigurationKeys.DistributedMode distributedMode, String taskId);
+
+    /** For AM, it will pre-check tony conf and inject some params. like horovod runtime will inject driver config into it. **/
```
Comment (oliverhu) [inline docs/proposals/horovod-on-tony.md:73]: `check` should not change things. Name it `validateAndUpdateConfig`?

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java [hunk 9]
```diff
@@ -426,7 +380,91 @@ private void skewAndHangIfTesting() {
...
   }
+
+  public ServerPort getRpcPort() {
+    return rpcPort;
+  }
+
+  public ServerPort getTbPort() {
+    return tbPort;
+  }
+
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java:394]: a map of ports? We have getRpcPort, getTBPort, getAmport and getMetricsRPCport here

File: tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java [hunk 10]
```diff
@@ -426,7 +380,91 @@ private void skewAndHangIfTesting() {
...
+  public int getAmPort() {
+    return amPort;
+  }
+
+  public int getMetricsRPCPort() {
+    return metricsRPCPort;
+  }
+
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java:410]: we have 4 getters for port.. shall we make this more generic?

File: docs/proposals/horovod-on-tony.md (hunk 7 above)
Comment (zuston) [reply docs/proposals/horovod-on-tony.md:70]: Got it

File: tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+import org.codehaus.jackson.map.ObjectMapper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.linkedin.tony.util.Utils;
+
+import static java.util.Objects.requireNonNull;
+
+public class HorovodDriver {
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:31]: Add some comments for HorovodDriver, not very intuitive to understand the role here. I know you had some content in proposal, but pls also some in the code. :)

File: tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+                        new TypeToken<List<SlotInfo>>() { }.getType());
+                return Pair.of(port, slotInfoList);
+            }
+        }
+        LOG.info("Still no starting horovod rendezvous server.");
+        return Pair.of(port, null);
+    }
+
+    private static boolean existPortFile(Path parentPath) throws IOException {
+        return getServerInfo(parentPath).getLeft() != -1 ? true : false;
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:170]: will there be a race condition on port allocation if there are a horovod job and a non-horovod job run in parallel?

File: tony-core/src/main/java/com/linkedin/tony/horovod/SlotInfo.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2021 LinkedIn Corporation. All rights reserved. Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.tony.horovod;
+
+public class SlotInfo {
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/horovod/SlotInfo.java:7]: would be useful to have some documentation for slotinfo here.. i know that is in Horovod documentation, but would be nice to have some info here.

File: tony-core/src/main/java/com/linkedin/tony/runtime/HorovodRuntime.java [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+            return null;
+        }
+
+        log.info("Starting Horovod worker, task id: " + taskId);
+        // when task role is worker, it will return horovod cluster spec.
+        HorovodClusterSpec spec = new HorovodClusterSpec(
+                workerSlotMetaInfo,
+                rendezvServerPort,
+                rendezvServerHost,
+                sameHostIndexCollection
```
Comment (oliverhu) [inline tony-core/src/main/java/com/linkedin/tony/runtime/HorovodRuntime.java:91]: indentation should be 4 spaces across the files

**Review by oliverhu — COMMENTED** [review]
left some comments, overall looks good! pls fix the ci

File: tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java (hunk 11 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:31]: Got it.

File: tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java (hunk 12 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:170]: This won't happen. The port files will be created in independent application folder.

File: tony-core/src/main/java/com/linkedin/tony/horovod/SlotInfo.java (hunk 13 above)
Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/horovod/SlotInfo.java:7]: Got it.

File: tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java (hunk 12 above)
Comment (oliverhu) [reply tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:177]: Not file conflict. I meant port conflict. We had this issue before that if you launch multiple TF jobs on the same node, there might be a race condition. The problem was, we reserve a port in Java for the job, and later release it before the job starts. However, the job might not start right away and would take a few minutes to set up. During that few minutes, another TF job could take up that port and causes port conflict.

Comment (zuston) [reply tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:177]: It won't happen. In horovod driver, It is Python process(built-in horovod_driver.py) that selects an available port, which is not TonY reserved. Therefore, port race condition will not occur.

Comment (oliverhu) [reply tony-core/src/main/java/com/linkedin/tony/horovod/HorovodDriver.java:177]: Got it, makes sense

**Comment by zuston** [comment]
Fix all. please review again @oliverhu

**Review by oliverhu — APPROVED** [review]

**Comment by oliverhu** [comment]
lgtm, thanks for the great work!!

**Pull request merged**